¡Hola, Andrea!

Mi nombre es Tonatiuh Cruz. Me complace revisar tu proyecto hoy.

Al identificar cualquier error inicialmente, simplemente los destacaré. Te animo a localizar y abordar los problemas de forma independiente como parte de tu preparación para un rol como data-scientist. En un entorno profesional, tu líder de equipo seguiría un enfoque similar. Si encuentras la tarea desafiante, proporcionaré una pista más específica en la próxima iteración.

Encontrarás mis comentarios a continuación - **por favor no los muevas, modifiques o elimines**.

Puedes encontrar mis comentarios en cajas verdes, amarillas o rojas como esta:

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Éxito. Todo está hecho correctamente.
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Observaciones. Algunas recomendaciones.
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Necesita corrección. El bloque requiere algunas correcciones. El trabajo no puede ser aceptado con comentarios en rojo.
</div>

Puedes responderme utilizando esto:

<div class="alert alert-block alert-info">


# PROYECTO SQL
El objetivo de este estudio es estudiar las reseñas de usuarios para libros así como la editorial con más libros con más de 50 páginas. También queremos identificar a los autores con calificaciones más altas y entender el comportamiento de los usuarios que hacen reseñas. 


In [1]:
# importar librerías


import pandas as pd
from sqlalchemy import create_engine


db_config = {'user': 'practicum_student',         # nombre de usuario
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # contraseña
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # puerto de conexión
             'db': 'data-analyst-final-project-db'}          # nombre de la base de datos

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})



## 1. Estudiar las tablas

In [4]:

tablas = ["books", "authors", "publishers", "ratings", "reviews"]

def tables(tabla, n=10):
    query=f"SELECT * FROM {tabla} LIMIT {n};"
    print(f"{tabla.upper()}")
    return pd.io.sql.read_sql(query, con = engine)

for table in tablas:
    display(tables(table))

BOOKS


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

AUTHORS


author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

PUBLISHERS


publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

RATINGS


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

REVIEWS


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

## 2. CONSULTAS SQL

## 2.1 Número de libros publicados después del 1 de enero de 2000

In [19]:


query = """
SELECT * 
FROM books
WHERE publication_date > '2000-01-01'
ORDER BY publication_date ASC;
"""
df_books = pd.io.sql.read_sql(query, con = engine)
display(df_books)


book_id  author_id                                              title  \
0         77        414  Angels Flight (Harry Bosch  #6; Harry Bosch Un...   
1        778        571  The Light Fantastic (Discworld  #2; Rincewind #2)   
2        736        231                                The Gods Themselves   
3        725        515                                           The Firm   
4        934         30          Traveling Mercies: Some Thoughts on Faith   
..       ...        ...                                                ...   
814      993         80                                              Women   
815      293         80                                         Ham on Rye   
816      445        612                                            Monster   
817      635        166                                  The Art of Loving   
818       43        377                         A Quick Bite (Argeneau #1)   

     num_pages publication_date  publisher_id  
0          454       2000-01-05           327  
1          277       2000-02-02           143  
2          288       2000-02-10           182  
3           76       2000-02-15             5  
4          275       2000-02-15            15  
..         ...              ...           ...  
814        291       2014-07-29            99  
815        288       2014-07-29            99  
816        281       2019-03-05            14  
817        192       2019-08-06           130  
818        360       2020-03-31            28  

[819 rows x 6 columns]

Existen 819 libros publicados a partir del 2000. 

## 2.2 Número de reseñas de usuarios y calificación promedio para cada libro

In [20]:
query = """

SELECT books.book_id,
       books.title,
       COUNT(DISTINCT reviews.review_id) AS review_count,
       ROUND(AVG(ratings.rating), 2)    AS avg_rating
FROM books
LEFT JOIN reviews ON books.book_id = reviews.book_id
LEFT JOIN ratings ON books.book_id = ratings.book_id
WHERE books.publication_date > DATE '2000-01-01'
GROUP BY books.book_id, books.title
ORDER BY avg_rating DESC, review_count DESC;

"""
df_rate_review = pd.io.sql.read_sql(query, con = engine)
display(df_rate_review)


book_id                                              title  review_count  \
0         17                      A Dirty Job (Grim Reaper  #1)             4   
1        444       Moneyball: The Art of Winning an Unfair Game             3   
2        553            School's Out—Forever (Maximum Ride  #2)             3   
3        182                                         Dead Souls             2   
4        516                       Plum Island (John Corey  #1)             2   
..       ...                                                ...           ...   
814      772                             The Kitchen God's Wife             3   
815      915  The World Is Flat: A Brief History of the Twen...             3   
816      202                                      Drowning Ruth             3   
817      371                                              Junky             2   
818      316                  His Excellency: George Washington             2   

     avg_rating  
0          5.00  
1          5.00  
2          5.00  
3          5.00  
4          5.00  
..          ...  
814        2.33  
815        2.25  
816        2.00  
817        2.00  
818        2.00  

[819 rows x 4 columns]

Entre los 819 libros, algunos alcanzaron calificaciones promedio altas (5.00) con varias reseñas, como A Dirty Job, Moneyball o School's Out—Forever. También se detectan títulos con calificaciones bajas (por ejemplo Drowning Ruth con 2.00). Todos estos son libros publicados a partir del 2000.

## 2.3 Identifica la editorial que ha publicado el mayor número de libros con más de 50 páginas (esto te ayudará a excluir folletos y publicaciones similares de tu análisis)

In [25]:
query = """


SELECT publishers.publisher_id,
       publishers.publisher,
       COUNT(books.book_id) AS total_books_over_50
FROM books
JOIN publishers ON books.publisher_id = publishers.publisher_id
WHERE books.num_pages > 50
    AND books.publication_date > DATE '2000-01-01'
GROUP BY publishers.publisher_id, publishers.publisher
ORDER BY total_books_over_50 DESC
LIMIT 20;

"""

over_fifty = pd.io.sql.read_sql(query, con = engine)
display(over_fifty)



publisher_id                      publisher  total_books_over_50
0            212                  Penguin Books                   35
1            116       Grand Central Publishing                   24
2            217               Penguin Classics                   22
3            309                        Vintage                   20
4             35                         Bantam                   19
5             33               Ballantine Books                   15
6             45                        Berkley                   14
7             83                          Delta                   13
8            129               Harper Perennial                   12
9            284             St. Martin's Press                   12
10           176      Little  Brown and Company                   11
11            46                  Berkley Books                   11
12           210                        Penguin                   10
13           143                    HarperTorch                   10
14           333      William Morrow Paperbacks                   10
15           274                    Simon Pulse                    9
16           329        Washington Square Press                    9
17            31                 Back Bay Books                    9
18           133                  HarperCollins                    9
19           243  Random House Trade Paperbacks                    9

La editorial penguin Books tiene la mayor cantidad de libros con más de 50 páginas (y publicados a partir del año 2000), con un total de 35 libros.

## 2.4 Identifica al autor que tiene la más alta calificación promedio del libro: mira solo los libros con al menos 50 calificaciones.

In [32]:
query= """
SELECT authors.author_id,
       authors.author,
       ROUND(AVG(ratings.rating), 2) AS avg_rating,
       COUNT(ratings.rating) AS total_ratings
FROM books
JOIN authors ON books.author_id = authors.author_id
JOIN ratings ON books.book_id = ratings.book_id
WHERE books.publication_date > DATE '2000-01-01'
GROUP BY authors.author_id, authors.author
HAVING COUNT(ratings.rating) >= 50
ORDER BY total_ratings DESC
LIMIT 20;
"""
fifty_ratings = pd.io.sql.read_sql(query, con = engine) 
display(fifty_ratings)

author_id                                             author  avg_rating  \
0         236                         J.K. Rowling/Mary GrandPré        4.29   
1         240                                     J.R.R. Tolkien        4.24   
2         554                                    Stephenie Meyer        3.66   
3         106                                          Dan Brown        3.74   
4         445                                    Nicholas Sparks        3.87   
5         235                                      J.D. Salinger        3.85   
6         498                                       Rick Riordan        4.13   
7         195            George Orwell/Boris Grabnar/Peter Škerl        3.73   
8         291                                       Jodi Picoult        3.95   
9         311                                     John Steinbeck        3.64   
10        302                                       John Grisham        3.97   
11        630  William Shakespeare/Paul Werstine/Barbara A. M...        3.79   
12        372                                         Lois Lowry        3.74   
13        257                                    James Patterson        3.86   
14        536                                    Sophie Kinsella        3.88   
15         39                                      Arthur Golden        4.11   
16        376                                  Louisa May Alcott        4.20   
17        402                  Markus Zusak/Cao Xuân Việt Khương        4.26   
18          3                                    Agatha Christie        4.28   
19        130                                     Diana Gabaldon        4.30   

    total_ratings  
0             232  
1             166  
2             160  
3             143  
4             104  
5              98  
6              84  
7              74  
8              74  
9              73  
10             70  
11             66  
12             65  
13             64  
14             57  
15             56  
16             54  
17             53  
18             53  
19             50

El autor con calificaciones más altas promedio y al menos 50 calificaciones es Diana Gabaldon con una calificación de 4.3 y 50 calificaciones. Es importante señalar que la base de datos presenta registros con hasta 232 calificaciones. Sin embargo, esta incluye a múltiples autores, como es el caso de J.K. Rowling/Mary GrandPré con 232 calificaciones y un rating de 4.29

## 2.5 Encuentra el número promedio de reseñas de texto entre los usuarios que calificaron más de 50 libros.

In [34]:

query = """
SELECT ROUND(AVG(user_reviews.total_reviews), 2) AS avg_reviews_among_active_users
FROM (
    SELECT ratings.username,
           COUNT(DISTINCT reviews.review_id) AS total_reviews
    FROM ratings
    LEFT JOIN reviews ON ratings.username = reviews.username
    GROUP BY ratings.username
    HAVING COUNT(ratings.rating_id) > 50
) AS user_reviews;


"""
active_users = pd.io.sql.read_sql(query, con = engine) 
display(active_users)



avg_reviews_among_active_users
0                           17.46

En el caso de usuarios que han calificado más de 50 libros, el promedio de reseñas es de 17 reseñas por usuario.

## 3. Conclusiones finales

-Desde 2000 se publicaron 819 libros en la base analizada.

-Las calificaciones son muy variadas: desde 5 muy bueno a valoraciones muy bajas.

-La editorial Penguin Books cuenta con el mayor número de publicaciones con más de 50 hojas posteriores al año 2000.

-Diana Gabaldon es la autora mejor valorada, seguida por Rowling y Christie.

-Los usuarios más activos aportan en promedio 17 reseñas cada uno.

<div class="alert alert-block alert-success">
<b>Paso 2</b> <a class="tocSkip"></a><br>
Has realizado un buen trabajo al formular una consulta SQL para contar con:
    
    - el número de libros publicados después de una fecha específica.
    - Calificación promedio
    - Editorial que ha publicado la mayor cantidad de libros
    - Autor con la calificación promedio más alta
    - Promedio de reseñas
    
Esto demuestra tu habilidad para manipular y consultar bases de datos de manera efectiva. Solamente para terminar de complementar te recomiendo incluir el número de libros evaluados por autor para proporcionar un contexto adicional sobre la consistencia de sus calificaciones.
</div>